pep-0705.rst

```diff
@@ -0,0 +1,246 @@
+
+    class Movie(TypedDict):
+        name: str
+        year: NotRequired[int | None]
+
+But suppose we have another type where year is required::
+
+    class MovieRecord(TypedDict):
+        name: str
+        year: int
+
+Attempting to pass a ``MovieRecord`` into ``movie_string`` results in the error (using mypy):
+
+.. code-block:: text
+
+    Argument 1 to "movie_string" has incompatible type "MovieRecord"; expected "Movie"
+
+This particular use case should be type-safe, but the type checker correctly stops the
+user from passing a ``MovieRecord`` into a ``Movie`` parameter in the general case, because
+the ``Movie`` class has mutator methods that could potentially allow the function to break
+the type constraints in ``MovieRecord`` (e.g. with ``movie["year"] = None`` or ``del movie["year"]``).
+The problem disappears if we don't have mutator methods in ``Movie``. This could be achieved by defining an immutable interface using a :pep:`544` :class:`~typing.Protocol`::
+
+    from typing import Literal, Protocol, overload
+
+    class Movie(Protocol):
+        @overload
+        def get(self, key: Literal["name"]) -> str: ...
+
+        @overload
+        def get(self, key: Literal["year"]) -> int | None: ...
+
+        @overload
+        def __getitem__(self, key: Literal["name"]) -> str: ...
+
+        @overload
+        def __getitem__(self, key: Literal["year"]) -> int | None: ...
+
+This is very repetitive, easy to get wrong, and is still missing important method definitions like ``__contains__()`` and ``keys()``.
+
+Rationale
+=========
+
+The proposed ``TypedMapping`` type allows a straightforward way of defining these types that should be familiar to existing users of ``TypedDict`` and support the cases exemplified above::
+
+    from typing import NotRequired, TypedMapping
+
+    class Movie(TypedMapping):
+        name: str
+        year: NotRequired[int | None]
+
+In addition to those benefits, by flagging arguments of a function as ``TypedMapping``, it makes explicit not just to typecheckers but also to users that the function is not going to modify its inputs, which is usually a desirable property of a function interface.
+Finally, this allows bringing the benefits of ``TypedDict`` to other mapping types that are unrelated to ``dict``.
+
+Specification
+=============
+
+The ``TypedMapping`` type is a protocol behaving almost identically to ``TypedDict``, except:
+
+* The runtime type of a ``TypedMapping`` object is not constrained to be a ``dict``
+* No mutator methods (``__setitem__``, ``__delitem__``, ``update``, etc.) will be generated
+* A class definition defines a ``TypedMapping`` protocol if and only if ``TypedMapping`` appears directly in its class bases
+* Subclasses can narrow field types, in the same manner as other protocols
+
+All current and future features of ``TypedDict`` are applicable to ``TypedMapping``, including class-based and alternative syntax, totality, and ``Required``/``NotRequired`` from :pep:`655`.
```

Comment:
```suggestion
All other current features of ``TypedDict`` are applicable to ``TypedMapping``, including class-based and alternative syntax, totality, and ``Required``/``NotRequired`` from :pep:`655`.
```

* Make @JelleZijlstra 's [comment](https://github.com/python/peps/pull/2997#discussion_r1096703542) removing "and future" into a suggestion
* Also explicitly clarify that it is all features except for those listed above

Reply:
This was obsoleted and de-facto implemented by further changes